lib_test/test_httpaf.ml

```diff
@@ -154,9 +198,121 @@ module Server_connection = struct
       `Close (next_read_operation t);
   ;;
 
+  let test_synchronous_error () =
+    let request_string = "/ GET HTTP/1.1\r\n\r\n" in
+    let writer_woken_up = ref false in
+    let t = create ~error_handler synchronous_raise in
+    Alcotest.(check (of_pp Write_operation.pp_hum)) "Writer is in a yield state"
+      `Yield (next_write_operation t);
+    yield_writer t (fun () -> writer_woken_up := true);
+    let c = read_string t request_string in
+    Alcotest.(check int) "read consumes all input"
+      (String.length request_string) c;
+    Alcotest.(check (of_pp Read_operation.pp_hum)) "Error shuts down the reader"
+      `Close (next_read_operation t);
+    Alcotest.(check bool) "Writer woken up"
+      true !writer_woken_up;
+    Alcotest.(check (option string)) "Error response written"
+      (Some "HTTP/1.1 500 Internal Server Error\r\n\r\ngot an error")
+      (next_write_operation t |> Write_operation.to_write_as_string)
+  ;;
+
+  let test_synchronous_error_asynchronous_handling () =
+    let request_string = "/ GET HTTP/1.1\r\n\r\n" in
+    let writer_woken_up = ref false in
+    let continue = ref (fun () -> ()) in
+    let error_handler ?request error start_response =
+      continue := (fun () ->
+        error_handler ?request error start_response)
+    in
+    let t = create ~error_handler synchronous_raise in
+    Alcotest.(check (of_pp Write_operation.pp_hum)) "Writer is in a yield state"
+      `Yield (next_write_operation t);
+    yield_writer t (fun () -> writer_woken_up := true);
+    let c = read_string t request_string in
+    Alcotest.(check int) "read consumes all input"
+      (String.length request_string) c;
+    Alcotest.(check (of_pp Write_operation.pp_hum)) "Writer is in a yield state"
+      `Yield (next_write_operation t);
+    !continue ();
+    Alcotest.(check (of_pp Read_operation.pp_hum)) "Error shuts down the reader"
```

Comment:
This assertion is true before the `!continue ();`, right? Or even before the `next_write_operation`. It might be more clear if it read:

- Writer is in a yield state
- continue
- Writer woken up

Reply:
When in asynchronous raise cases (not asynchronous error generation), the reader will be closed only after `!continue ()` is called. So the read operation check should remain. However, it could be put below the "writer woken up test", if you think that's clearer.

Reply:
In that one since the order is significant I don't think I have a strong preference